src/Client.php

```diff
@@ -149,6 +149,28 @@ public function sendRequest(RequestInterface $request) : ResponseInterface
         return $response;
     }
 
+    /**
+     * Manipulate $curlOptions array
+     *
+     * @param array $curlOptions
+     *
+     * @return void
+     */
+    public function setCurlOptions(array $curlOptions) : void
+    {
+        $this->curlOptions = $this->curlOptions + $curlOptions;
```

Comment:
Hey @peter279k here's a REPL showing array + array https://repl.it/repls/PoliticalEnormousCheckpoint

I'm not 100% sure that this method will update & overwrite the `$this->curlOptions` unless it's blank

See a newer PHP interpreter example here http://sandbox.onlinephpfunctions.com/code/0e6979a720b116ec6874fce807861209aa0a873b

Reply:
Interested.

The CURL option array stores about the `CURL_OPT*` consts and I think this will not do plus operator with the two ordered array.

The `array_merge` is not good. Once the two arrays are merged, the arrays will indexed again.

If we want to prevent this, I think we only use the collections or using the loop to traverse array to add/override array manually.

Reply:
what do you have against `array_map` & `array_filter`?

Is it a memory concern?
perhaps you didn't know about them, or how they might apply here?

The code you've shared is very generic, probably quite efficient, but also doesn't communicate what it's doing or provide safety. The `array_` functions were made for this. 

Personally I'd have something like this http://sandbox.onlinephpfunctions.com/code/05dd69f787c0125c6b6f90677c2ca80e1d2e3e11

Also I'm not sure why you chose `func_get_args`, over variadic functions http://php.net/manual/en/migration56.new-features.php#migration56.new-features.variadics if you wish to stick with your design.

Reply:
@Lewiscowles1986, thank you for your reply.

- Using the `...` syntax that I forgot and it's useful for multiple arguments.
- I think your set/get options approach is not the good way and I also figure out we can use the `ArrayIterator` to manipulate options.

The related sample code is as follows:

```php
$arrObj = new ArrayObject([CURLOPT_AUTOREFERER => true, CURLOPT_URL => 'http://url.com']);

$arrObj = $arrObj->getIterator();

$arrObj->offsetSet(CURLOPT_BINARYTRANSFER, true);

var_dump(iterator_to_array($arrObj));
```

And we can also use the `ArrayIterator::offsetExists` to check whether specific cURL option is existed in option array list.

Reply:
What was it about that method I shared you didn't like?

Reply:
Hi @Lewiscowles1986, I prefer using `ArrayList` style to manipulate the cURL options.

But as @fenric says, we have to find a way to be compatible with the `PSR-18` rule.